gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala

```diff
@@ -86,6 +98,261 @@ class GcsApi private[gcs] (runHttp: Request[IO] => Resource[IO, Response[IO]]) {
+    }
+  }
+
+  /**
+    * Upload a chunk of bytes to an ongoing GCS resumable upload.
+    *
+    * @see https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#upload-resumable
+    *
+    * @param bucket the GCS bucket to upload the data into
+    * @param uploadToken the unique ID of the previously-initialized resumable upload
+    * @param rangeStart the (zero-indexed) position of the first byte of `chunkData` within
+    *                   the source file being uploaded
+    * @param chunkData chunk of bytes to upload
+    * @return either the number of bytes stored by GCS so far (signalling that the upload is not complete),
+    *         or Unit when all bytes have been received and GCS signals the upload is done
+    */
+  def uploadChunk(
+    bucket: String,
+    uploadToken: String,
+    rangeStart: Long,
+    chunkData: Stream[IO, Byte],
+    bytesPerUploadRequest: Int = MaxBytesPerUploadRequest
+  ): IO[Either[Long, Unit]] = {
+    val chunks = chunkData.chunkN(bytesPerUploadRequest)
+
+    chunks.zipWithIndex.evalMap {
+      case (chunk, index) =>
+        val chunkSize = chunk.size
+        val chunkStart = bytesPerUploadRequest * index
+
+        val gcsReq = Request[IO](
+          method = Method.PUT,
+          uri = baseGcsUploadUri(bucket, "resumable")
+            .withQueryParam("upload_id", uploadToken),
+          headers = Headers.of(
+            `Content-Length`.unsafeFromLong(chunkSize.toLong),
+            `Content-Range`(chunkStart, chunkStart + chunkSize - 1)
+          ),
+          body = Stream.chunk(chunk)
+        )
+
+        runHttp(gcsReq).use { response =>
+          if (response.status.code == 308) {
+            val bytesReceived = for {
+              range <- response.headers.get(`Content-Range`)
+              rangeEnd <- range.range.second
+            } yield {
+              rangeEnd
+            }
+            IO.pure(Left(bytesReceived.getOrElse(rangeStart + chunkSize)))
+          } else if (response.status.isSuccess) {
+            IO.pure(Right(()))
+          } else {
+            IO.raiseError(
+              new RuntimeException(
+                s"Failed to upload chunk with upload token: $uploadToken"
+              )
+            )
+          }
+        }
+    }.compile.lastOrError
+  }
+
+  /**
+    * Delete an object in GCS.
+    *
+    * TODO: Make this succeed if the object is already deleted, if possible.
+    *
+    * @param bucket name of the bucket containing the object to delete
+    * @param path path within `bucket` pointing to the object to delete
+    */
+  def deleteObject(bucket: String, path: String): IO[Unit] = {
+    val gcsUri = baseGcsUri(bucket, path)
+    val gcsReq = Request[IO](method = Method.DELETE, uri = gcsUri)
+
+    runHttp(gcsReq).use { response =>
+      if (response.status.isSuccess) {
+        IO.unit
+      } else {
+        IO.raiseError(new RuntimeException(s"Failed to delete object $path in $bucket"))
```

Comment:
Can you include the status code in this error?

Reply:
Actually, even better:

Can you pull the logic from the `else` block [here](https://github.com/broadinstitute/monster-storage-libs/pull/4/files#diff-4f0d9ee9f799079fd23a8eff3429a161R90) into a private method?
```scala
def reportError(failedResponse: Response[IO]): IO[Unit] = // The logic from that else block
```
then have all the new methods call that in the failure case? That way we get uniform error reporting.